src/arangodb.c

```diff
@@ -0,0 +1,1201 @@
+//
+static int arangodb_add_cluster(oconfig_item_t * cluster_config) {
+  int ret_val;
+  cluster_t * cluster;
+  
+  ret_val = 0;
+  
+  cluster = new_cluster();
+  if (NULL != cluster) {
+    ret_val = cluster_set_options(cluster_config, cluster);
+
+    if ( 0 == ret_val) {
+      start_curl_session(&cluster->curlinfo);
+      
+      if ( NULL == cluster->instance_name) {
+        cluster->instance_name = strdup("default");
+      } // if
+
+      // trying to make a unique name if several clusters
+      cluster->registered_name = ssnprintf_alloc("arangodb-%s-%s-%s", cluster->instance_name,
+                                                 cluster->endpoint.host, cluster->endpoint.port);
+
+      DEBUG("arangodb plugin: Registering new read callback: %s", cluster->registered_name);
+
+      plugin_register_complex_read(/* group = */ NULL, cluster->registered_name,
+                                   cluster_read, 0, //interval,
+                                   &(user_data_t){
+                                     .data = cluster,
+                                     .free_func = free_cluster,
+                                    });
+    } // if
+  } else {
+    ret_val = -1;
+  } // else
+
+  return ret_val;
+  
+} // arangodb_add_cluster
+
+
+//
+// Parse cluster options, set cluster_t object accordingly
+//  (get endpoint today, maybe other options in future, like user & password)
+//
+static int cluster_set_options(oconfig_item_t * cluster_config,
+                                cluster_t * cluster) {
+  int loop, ret_val;
+  oconfig_item_t * child;
+  
+  ret_val = 0;
+  
+  if (1 == cluster_config->values_num) {
+    ret_val = cf_util_get_string(cluster_config, &cluster->instance_name);
+  } // if
+  
+  for (loop=0, child=cluster_config->children;
+       loop<cluster_config->children_num && 0 == ret_val;
+       ++loop, ++child) {
+    if (0 == strcasecmp("endpoint", child->key)) {
+      ret_val = cf_util_get_string(child, &cluster->endpoint.given);
+    } // if
+  } // for
+
+  if (0 == ret_val) {
+    ret_val = cluster_verify_options(cluster);
+  } // if
+
+  return ret_val;
+
+} // cluster_set_options
+
+
+//
+// review all cluster options as a "set" to see if valid
+//  (later verify all things needed for ssl and such)
+static int cluster_verify_options(cluster_t * cluster) {
+  bool good = true;
+
+  if (good) {
+    // transform arangodb endpoint syntax to http stuff
+    good = endpoint_parse(&cluster->endpoint);
+  } // if
+
+  return (good ? 0 : -1);
+  
+} // cluster_verify_options
+
+
+//
+// collectd callback every "interval" to poll cluster,
+//  used to initiate poll to get cluster configuration
+//
+static int cluster_read(user_data_t *ud) {
+  int ret_val;
+  cluster_t * cluster;
+  
+  ret_val = 0;
+
+  if ( NULL != ud && NULL != ud->data) {
+    char * role = NULL;
+    cluster = (cluster_t *)ud->data;
+
+    role = arangodb_get_role(&cluster->curlinfo, &cluster->endpoint);
```

Comment:
In your implementation, if call to `arangodb_get_role` fails, code mades the wrong decisions, I think.